### PR TITLE
[EVM] Support arbitrary metadata

### PR DESCRIPTION
**Pull request by PavelKopyl**
This patch:

-  Adds 'INVALID' instruction at the end of .text section
    This is to mimic solc behaviour.
- Add support of metadata
    It adds the LLVMAddMetadata API function, shared by both
    EraVM and EVM targets, which inserts a .xvm-metadata section into
    an ELF file. The LLVMAssembleEVM function retrieves the metadata
    section (if present) from the first object file and appends it
    to the .text section of the resulting ELF file.

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp [hunk 1]
```diff
@@ -315,6 +315,22 @@ void EVMAsmPrinter::emitWideRelocatableSymbol(const MachineInstr *MI) {
 }
 
 void EVMAsmPrinter::emitEndOfAsmFile(Module &) {
+  // Construct a local MCSubtargetInfo here.
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: I'd moved it to emission and wrote something alike:
// Build a local MCSubtargetInfo as MachineFunction info is unavailable at this global phase.
// It's also not needed to emit INVALID instruction at the end of the .text section.

**Review by akiramenai — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: >I'd moved it to emission and wrote something alike:

Sorry, I didn't get. Do you mean to move the creation of local MCSubtargetInfo and emission of the instruction to EVMTargetStreamer?

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 2]
```diff
@@ -258,7 +258,7 @@ static std::string createEraVMRelLinkerScript(
 ///
 ///       /* Add padding before the metadata. Here metadata size is 32 bytes */
 ///       . = ((((. + 32) >> 5) | 1 ) << 5) - 32;
-///       *(.eravm-metadata)
+///       *(.xvm-metadata)
```
Comment (hedgar2017) [inline lld/lld-c/LLDAsLibraryC.cpp]: Do we need to introduce this `xvm` here?
Can be misleading. Can we just call it `metadata` or in another LLVM-ish way?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 2 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp]: Sure, renamed to `metadata`.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Done.

**Review by akiramenai — APPROVED** [review]

**Review by hedgar2017 — APPROVED** [review]
